mst-form-type documentation in markdown as we talked in email (#2168)
mobxjs/mobx-state-tree: docs/libraries/mst-form-type.md

# MST Form Type

> npm: https://www.npmjs.com/package/mst-form-type

Reviewer: ```suggestion
Find the `mst-form-type` library on npm: https://www.npmjs.com/package/mst-form-type.
```

**suggestion:** I'd prefer to lead with a full sentence if you don't mind.

## Introduction

> Librarys like Ant Design has a build-in Form component that can hold fields status and validation rules. This component houses its own store to sustain internally related field statuses and can be conveniently used independently. integrating Form component with MobX State Tree models can pose significant challenges as business logic grows intricately complex.

Reviewer: ```suggestion
Libraries like [Ant Design](https://ant.design/) have a built-in `Form` component that holds field status and validation rules.  Integrating a `Form` component with MobX State Tree models can pose significant challenges as business logic become more complex.
```

**suggestion:** I'd like to fix some spelling and grammar errors here, and also clarify some of the sentences.

> Librarys like Ant Design has a build-in Form component that can hold fields status and validation rules. This component houses its own store to sustain internally related field statuses and can be conveniently used independently. integrating Form component with MobX State Tree models can pose significant challenges as business logic grows intricately complex.

Reviewer: **question:** can you please list a few examples of those challenges? It would be great to articulate more about why the `Form` component is hard to use with MST models.

> The `mst-form-type` library comes into play. It provides the field management feature of Ant Design Form component in Mobx State Tree model way. Users can only use the UI of a component library, but keep logic inside your Mobx State Tree model context, instead of syncing status changes between Form component and model.

Reviewer: ```suggestion
That's where a solution like `mst-form-type` library comes into play. It models the Ant design field management like a conventional MobX-State-Tree type definition. Users can still use the UI of a component library and keep logic inside MobX-State-Tree, instead of syncing status changes between `Form` component and model.
```

**suggestion:** I reworded a few things and fixed spelling/grammar errors here.

Please note: The `mst-form-type` library primarily provides model types for the form structure. It does not encompass business logic related to field interactions.

## Setup

Setting up and utilizing the `mst-form-type` library is straightforward:

1. Install the `mst-form-type` library via npm:

```sh
npm install mst-form-type
```

2. Ensure you have `mobx-state-tree ^5.0.0` installed:

```sh
npm install mobx-state-tree@^5.0.0
```

3. Declare a form schema and create the form model, subsequently, all the relevant field models would then be created in a form model. An example schema look like this:

> ```typescript

Reviewer: **suggestion:** can you show a full example, including all the imports, this schema definition, and then how the library would use it? I see something like that on npm:

```js

// render dynamic fields
model.dynamicForm['price'].fields.map(fields => ({ ... }))

// field action
model.dynamicForm.onAdd('price') // this will use field default value in schema
model.dynamicForm.onRemove('price', fields.id)

// form action
model.dynamicForm.submit()
model.dynamicForm.reset()
model.dynamicForm.clear('price') // this will clear all dynamic fields, including default ones
```

const dynamicForm: FormSchema = {
  static: [
    {
      id: 'name',
      default: '',
      validator: 'required',
    },
    {
      id: 'des',
      default: '',
    },
  ],
  dynamic: [
    {
      id: 'price', // group id
      limit: 100,
      schema: [
        {
          id: 'itemName',
          default: '',
          validator: 'required',
        },
        {
          id: 'itemPrice',
          default: 10,
        },
      ],
      default: [
        {
          itemName: 'itemName1',
          itemPrice: 5,
        },
        {
          itemName: 'itemName2',
          itemPrice: 20,
        },
      ],
      onAdd: i => {
        console.log('add', i)
      },
      onRemove: i => {
        console.log('remove', i)
      },
      onEdit: key => {
        console.log('edit', key)
      },
    },
  ],
}

> ```

Reviewer: **suggestion:** can you go over the different types of fields? I'd love to pull more of the docs from [npm](https://www.npmjs.com/package/mst-form-type) into this if possible.

Maybe create an example of a form you've used in the past, and explain the use cases for different syntax. It's OK if you want to just cover the basic fields, as well.

One way or another, it would be awesome to get a more realistic example here.

## How It Works

> The `mst-form-type` library exports a function, which creates a form model based on the base form model underhood using a form schema and an optional name. Each form type model implements its own type, and every attribute and method is independent. The user can conveniently interact with the form field via form props, or apply the form model action to directly get or set field values.

Reviewer: ```suggestion
The `mst-form-type` library exports a function which creates a form model (based on the base form model under the hood) using the provided schema and an optional name. Each form type model implements its own type, and every attribute and method is independent. The user can conveniently interact with the form field via form props, or apply the form model actions to directly get or set field values.
```

**suggestion:** fixed some spelling and grammar errors.

> The `mst-form-type` library exports a function, which creates a form model based on the base form model underhood using a form schema and an optional name. Each form type model implements its own type, and every attribute and method is independent. The user can conveniently interact with the form field via form props, or apply the form model action to directly get or set field values.

Reviewer: **question:** what does it mean when you say "every attribute and method is independent"?

> ## Architecture

Reviewer: **suggestion:** can you add more detail to this section about how these three entities work together with the schema to become what the user will end up working with?
